[Wu Changjun] iP

Ugholaf opened the pull request:
# Duke
> "Your mind is for having ideas, not holding them." - David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~  SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/Ugholaf/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon!)
- [ ] Reminders (coming soon!)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
``` Java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

Ugholaf committed:
Add echo and exit

Ugholaf committed:
Include add and list feature

Ugholaf committed:
Create Task class

Ugholaf committed:
Add mark and unmark feature

Ugholaf committed:
Refactor executeCommand method

Ugholaf committed:
Create Todo, Event and Deadline class

Ugholaf committed:
Add support for tracking Todos, Events and Deadlines

Ugholaf committed:
Create DukeException class

Ugholaf committed:
Use DukeException to handle errors

Ugholaf committed:
Add delete tasks feature

Ugholaf committed:
Update error message

Ugholaf committed:
Update test files

Ugholaf committed:
Add TaskType Enum

Ugholaf committed:
Refactor methods to add new tasks

Ugholaf committed:
Add Storage class and save tasks to disk

Ugholaf committed:
Merge branch 'branch-Level-7'

DanielLimWeiEn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,230 @@
+
+        storage.appendTaskToFile(task);
+
+        String addTaskMessage =  "Got it. I've added this task:\n" +
```
Would it be a good idea to extract out the printing of the messages to a separate class, perhaps a Ui class? since the printing of messages would fit in a Ui related class.

DanielLimWeiEn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,230 @@
+        }
+    }
+
+    private static void deleteTaskCommand(String[] inputs) throws DukeException {
```
I like your naming of methods 👍. The names are very descriptive and very clearly convey what the methods are doing.

DanielLimWeiEn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,230 @@
+            Task task = taskList.get(taskIndex);
+            task.markAsDone();
+            storage.writeAllTasksToFile(taskList);
+
```
I'm not sure I like how the code here aims to do multiple things all at once. From carrying out the task to printing to the user interface. Do you think it is a good idea to maybe extract some of similar logic to separate classes?

DanielLimWeiEn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,230 @@
+
+        try {
+            switch (inputs[0]) {
+                // List out all tasks
```
I think that the switch case here isn't following the coding standard they've given to us namely the standard below with the indentation.

```java
switch {
case 1:
    // ...
case 2:
    // ...
default:
    break;
}
```

DanielLimWeiEn reviewed (commented):
Overall looks pretty good 👍. I like your naming of methods and variables as it is very obvious what the methods and variables are for. There were however some areas that could be done better to make the code better, such as extracting out some functionality.

Ugholaf committed:
Add Dates and Times to tasks

Ugholaf committed:
Merge branch 'branch-Level-8'

Ugholaf committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Ugholaf committed:
Create TextUi, Parser, TaskList and Command classes

Ugholaf committed:
Divide classes into packages

Ugholaf committed:
Add JUnit Tests

Ugholaf committed:
Fix mainClassName

Ugholaf committed:
Add JavaDoc comments

Ugholaf committed:
Adhere to coding standard

Ugholaf committed:
Add FindCommand for user to search task by keyword

Ugholaf committed:
Merge branch 'branch-A-JavaDoc'

Ugholaf committed:
Merge branch 'branch-A-CodingStandard'

Ugholaf committed:
Fix error in reading user input

Ugholaf committed:
Resolve merge conflict

Ugholaf committed:
Change the access method of prepareFindCommand

Ugholaf committed:
Refactor ListCommand and add JavaDoc

Ugholaf committed:
Standardize naming for taskList

Ugholaf committed:
Add JavaDoc for FindCommand

Ugholaf committed:
Add checkstyle and fix checkstyle errors

Ugholaf committed:
Merge branch 'branch-A-CheckStyle'

Ugholaf committed:
Add GUI

Ugholaf committed:
Merge branch 'branch-Level-10'

Ugholaf committed:
Add assertions

Use assertions to verify the correctness of various Task and Duke classes.

All tasks should have a description.
The getResponse method in Duke should never get an empty input.

Add assertions to Task constructors to ensure the description is not empty.
Add assertion to getResponse method in Duke to ensure the input is not empty.

This will ensure the classes behave as expected during runtime.

Ugholaf committed:
Minimize code duplication in Parser class

The prepareMarkCommand and prepareUnmarkCommand methods have very similar code and logic.

Remove code duplication to improve the code quality.

To remove code duplication, extract out the duplicated code blocks into separate method.

Ugholaf committed:
Use streams for findTask method

The findTask method in TaskList class is currently done by iterating through for loops to find the matching tasks.

Use streams and filter method to filter out the matching task.

Ugholaf committed:
Merge pull request #2 from Ugholaf/branch-A-Assertions

Add assertions

Ugholaf committed:
Merge pull request #4 from Ugholaf/branch-A-Streams

Use streams for findTask method

Ugholaf committed:
Merge pull request #3 from Ugholaf/branch-A-CodeQuality

Minimize code duplication in Parser class

Ugholaf committed:
Set up CI

Ugholaf committed:
Fix code style error

Ugholaf committed:
Add friendlier and shorter command syntax

Ugholaf committed:
Merge branch 'branch-C-FriendlierSyntax'

Ugholaf committed:
Add border radius to text box

Ugholaf committed:
Add colour to GUI

Ugholaf committed:
Add duke image and title to the title bar

Ugholaf committed:
Fix display error

Ugholaf committed:
Add product screenshot

Ugholaf committed:
Add a User Guide

Ugholaf committed:
Update User Guide

Ugholaf committed:
Fix GUI display error

Ugholaf committed:
Merge branch 'master' of https://github.com/Ugholaf/ip

Ugholaf committed:
Update mainClassName in build.gradle

Ugholaf committed:
Fix GUI text overflow error

Ugholaf committed:
Fix User Guide format

Ugholaf committed:
Fix header comments
